Add missing tiles to arrakis.yaml and fixed the map import

abcdefg30 opened the pull request:
My take on #12244.
Fixes #12202.

abcdefg30 committed:
Let the map importer account for duplicate tiles

abcdefg30 committed:
Add missing templates to arrakis.yaml

CH4Code commented:
All right, finaly someone with the "right" approach :)

ltem commented:
I converted all the multiplayer maps and checking them right now.

The only error I found so far is one missing/wrong tile.
Map: 2PLAY13
Tileset: Bloxwast.r8

| Original | OpenRA|
|:-------:|:-------:|
|![image](https://cloud.githubusercontent.com/assets/20945684/23074766/46c9a82a-f53a-11e6-8910-b745f1409576.png) | ![image](https://cloud.githubusercontent.com/assets/20945684/23074773/4ce5815c-f53a-11e6-85f0-5c7c52bcb306.png) |

The tile id from the right picture is 65 but I didn't found the correct tile in OpenRA.

ltem commented:
So I finally tested the mission maps now and I found only two tile errors.

Map: O7V1
Tileset: bloxwast

| Original | OpenRA |
|:------:|:------:|
| ![image](https://cloud.githubusercontent.com/assets/20945684/23101347/b20003ac-f691-11e6-86b9-02e48d5c96d9.png) | ![image](https://cloud.githubusercontent.com/assets/20945684/23101348/b5a0dbbc-f691-11e6-96e3-f87770103732.png) |
| | ![conclusion](https://cloud.githubusercontent.com/assets/20945684/23101571/f0369bfe-f696-11e6-8441-23bb058c1514.png) |

-------------------------------

__Additional notes:__
I don't know if the following points are in scope of the importer but I stumbled over them during the review process
- __Note 1:__ Some buildings (Corrino?) won't be imported correctly.

| Original | OpenRA |
|:------:|:------:|
| ![image](https://cloud.githubusercontent.com/assets/20945684/23101367/303b18c4-f692-11e6-9a55-3488e200dc3a.png) | ![image](https://cloud.githubusercontent.com/assets/20945684/23101372/39a67c28-f692-11e6-92da-6090075858ba.png)  |
| ![image](https://cloud.githubusercontent.com/assets/20945684/23101373/47ba96be-f692-11e6-8979-021d6f249175.png) | ![image](https://cloud.githubusercontent.com/assets/20945684/23101377/4f207248-f692-11e6-92e9-b480e4bd7683.png) |


- __Note 2:__ If a map has some kind of "padding" the importer will crash


|![scali](https://cloud.githubusercontent.com/assets/20945684/23101384/9af61b64-f692-11e6-83a1-33742726f4af.png)|
|:------------:|
|Map with black padding|

```
$ ./OpenRA.Utility.exe d2k --import-d2k-map missionMaps/A2V2.MAP bloxtree.r8
System.IndexOutOfRangeException: Array index is out of range.
  at OpenRA.CellLayer`1[OpenRA.TerrainTile].set_Item (CPos cell, TerrainTile value) [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.D2k.UtilityCommands.D2kMapImporter.FillMap () [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.D2k.UtilityCommands.D2kMapImporter..ctor (System.String filename, System.String tileset, OpenRA.Ruleset rules) [0x00000] in <filename unknown>:0
```

@pchote mentioned that a fix for such maps, could cause bogus imports for others. 
This only effects the first 4 Missions of every faction (A1V1, A1V2, A2V1,....), which are already imported by @CH4Code (https://github.com/CH4Code/D2k-Campaign-Resources), so ¯\\\_(ツ)\_/¯

MustaphaTR commented:
Side specific structure images in OpenRA are set according to side itself. If side is Harkonnen and not Corrino it is normal that it use Harkonnen Palace. Importer doesn't create players afaik, unless this PR added so.

Corrino, Mercenaries, Fremen and Smugglers are valid sides on OpenRA.

CH4Code commented:
I originaly didn't check the multiplayer maps for tile errors, since those were allready (manually) fixed and done (in the official map pool). So abcdefg30's corrected importer based on my (not meant to be overarching) fixes should only work 100% in case of missions.
@Item: The 1st 4 missions in my repo should not **all** display correctly in vanilla OpenRA, since they use a custom tileset (arrakis.yaml). Whould be surprised if they did.

CH4Code commented:
I could extend the custom tileset, which abcdefg30 then could translate into c#/map importer, fixing all multiplayer maps. But I will **only** do this if it is requested by abcdefg30 or pchote. It's a ton of work.

MustaphaTR commented:
Those black area in maps added on Gruntmods edition and doesn't exist in either original 1.2 or 1.6. It is added for Hi-Res Patch or they was crashing.

abcdefg30 commented:
@ltem thanks for going through all the missions again.

So as far as I understand, this only seems to fix the tileset for missions.
What about taking this (after I fixed O7V1) and adding the required new tiles as needed when importing the multiplayer maps? (So this PR doesn't get any larger.)

MustaphaTR commented on mods/d2k/tilesets/arrakis.yaml:
```diff
@@ -5671,3 +5671,381 @@ Templates:
+			0: Rough
+			1: Rough
+	Template@1084:
+		Id: 10309
```
This and 4 above don't have matching @ part and Id:

abcdefg30 replied:
Ops, I forgot to rename those. Thanks.

CH4Code commented:
@abcdefg30 seems fine, #12788 is for multiplayer maps and could have it's own pull request. Having the converter fixed for Missions will greatly benefit the campaign ticket (#9287).

ltem commented:
@abcdefg30 I'm totally fine with that

Most of the multiplayer tile errors seem to be caused by bloxxmas, which wasn't used for the missions, so it seems fitting to fix this with an other PR.

abcdefg30 commented:
Updated and fixed the remaining two tiles.

ltem commented:
O7V1 is fixed so :+1: from me

CH4Code commented:
Hi, went through the bugs in the MP maps as well. Those additional tiles (no duplicates) will fix importing all MP maps. 
Also I added a "Bridge" Category for all the bridge tiles. On 4P10, 6P4, 6P5 however, the bridges are partially only passable by infantry(due to other predefined tiles) - but manual fix in like 1 minute with the new tiles.
@abcdefg30 just add "Bridge" category and add these lines to your arrrakis.yaml:
[http://pastebin.com/C6xzM4xZ](http://pastebin.com/C6xzM4xZ) (valid for one week)
This will finalize our map-import-effort I hope :)

CH4Code committed:
Add the missing d2k tiles for the MP maps

abcdefg30 commented:
Thanks. Added it.

reaperrr commented on mods/d2k/tilesets/arrakis.yaml:
```diff
@@ -6049,3 +6049,236 @@ Templates:
+		Images: BLOXXMAS.R8
+		Frames: 345, 346
+		Size: 2,1
+		Category: Bridge
```
You need to add this category to `EditorTemplateOrder` to make them available in the editor (I suggest placing it before `Unidentified`).

reaperrr commented:
In addition to the category fix, the destroyed north/south bridge ends are currently missing.

I've created yaml definitions here:
https://gist.github.com/reaperrr/054c08bb66e6cbd8fda889e3d434d521

@Item @CH4Code can you confirm that they look as they should and have correct (im)passability?

CH4Code commented:
> In addition to the category fix, the destroyed north/south bridge ends are currently missing.

Yes and no. Look at Template 513 and 514 in rock detail (where the rest of the bridges are). If you copy/paste parts of it you can construct the big destroyed north/south ends. Thats why the map is generated correctly without them "missing", as is. 

The openRA devs originaly simply misinterpreted the BLOX* tile file, i guess (which is't hard at all :D ). There, the bridges are dispalyed in a "compressed" format.

> can you confirm that they look as they should and have correct (im)passability?

They are nor passable at all they way you defined them. Judging from  513 and 514 the centermost bridge tile should be "Rough"(passable for inf), the other "Cliff". 

@reaperrr I don't think it is needed to fix importing, though.
Of course detatched ends have better useability in map editor :)

ltem commented:
Uhh, I derped a bit, I should have seen that.

In the original tileset, the bridges are smaller.
![image](https://cloud.githubusercontent.com/assets/20945684/23381089/7f9455be-fd3d-11e6-8d8e-0b2a2db7b109.png)
Source: http://old.d2kplus.com/wiki/index.php?title=File:Bloxxmas.png


Infantry can pass the bridge "slopes" (but not vehicles), the cursor gets grey so it should be `Rough`
![image](https://cloud.githubusercontent.com/assets/20945684/23382578/fe0e1394-fd42-11e6-94f4-ce445905c8da.png)
![image](https://cloud.githubusercontent.com/assets/20945684/23383066/febebada-fd44-11e6-929c-33e6d4b0c1bd.png)


I would move the tiles 513, 514, 515 as well to the `Bridge` category.
![image](https://cloud.githubusercontent.com/assets/20945684/23382024/0022a8cc-fd41-11e6-98e9-f589afca64aa.png)

CH4Code commented:
@Item I think those are map editor issues and should get their own PR later.

CH4Code commented:
> Yeah "compressed" sounds very good. I don't know what is to prefer for map makers.

Smaller, but identifyable "building blocks" imho. But, if we remove/reorganize existing tiles alltogether, that surely will cause problems for existing maps.

CH4Code commented:
Nice :) Found the last fix :) Replace 
513    4: Rough -> 4: Transition    and  
514    4: Rough -> 4: Transition.
Then all passability issues are auto-fixed :)

ltem commented:
I agree with you  about the smaller building blocks and as you already said before (^^ I tried to :+1: it up but it was already gone) this should be done in an extra PR.

abcdefg30 commented:
Updated. Thanks @reaperrr (and @CH4Code and @ltem for constantly rechecking things ^^).

CH4Code commented:
@abcdefg30: you'll need to incorporate my comments on 513, 514, 515 and my comment on reaperrs added tiles, too, or well get passability issues needing post-map-generation manual fixes.

CH4Code commented on mods/d2k/tilesets/arrakis.yaml:
```diff
@@ -5671,3 +5671,650 @@ Templates:
+		Category: Bridge
+		Tiles:
+			0: Cliff
+			1: Cliff
```
1: Rough

CH4Code commented on mods/d2k/tilesets/arrakis.yaml:
```diff
@@ -5671,3 +5671,650 @@ Templates:
+		Size: 2,1
+		Category: Bridge
+		Tiles:
+			0: Cliff
```
1: Rough

CH4Code commented on mods/d2k/tilesets/arrakis.yaml:
```diff
@@ -5671,3 +5671,650 @@ Templates:
+		Size: 2,1
+		Category: Bridge
+		Tiles:
+			0: Cliff
```
0: Rough

CH4Code commented on mods/d2k/tilesets/arrakis.yaml:
```diff
@@ -5671,3 +5671,650 @@ Templates:
+		Size: 2,1
+		Category: Bridge
+		Tiles:
+			0: Cliff
```
0: Rough

CH4Code commented on mods/d2k/tilesets/arrakis.yaml:
```diff
@@ -5671,3 +5671,650 @@ Templates:
 		Category: Sand-Detail
 		Tiles:
 			0: Rough
```
@ Id: 513
Tiles:
     4: Transition

@ Id: 514
Tiles:
    4: Transition

@ Id: 515
Tiles:
			1: Cliff
			2: Cliff

CH4Code reviewed (changes requested):
Change "Tiles:" to fix remaining passability issues.

CH4Code commented:
Also ..... just some cosmetics : could you move` Template@1108:` to be before `Template@1085: `?
Has nice (originally intended, messed up ^^) effect on tile order in editor:
![grafik](https://cloud.githubusercontent.com/assets/16046997/23461907/195d8a36-fe8c-11e6-94bc-ddd282b116e6.png)

abcdefg30 commented:
Sorry for the delay, updated.

ltem commented:
The end is near :D

I went through all the multiplayer maps and the only tile error if found is Error 3 of my comment https://github.com/OpenRA/OpenRA/pull/12784#issuecomment-280706706. It still exits (Map: [6PLAY3.MAP.zip](https://github.com/OpenRA/OpenRA/files/819406/6PLAY3.MAP.zip) Tileset: bloxwast.r8)

So after Error 3 is fixed my :+1: valid again

CH4Code commented:
Yep, those two errors are C# solvable only.  👍, too, after this :)

abcdefg30 commented:
Sorry for the late update. Finally fixed it. This should be good to go now.

CH4Code commented:
Erm, can't make out any changes since last time o.O

ltem reviewed (approved):
6PLAY3 fixed

CH4Code reviewed (approved):
That last change finally fixed all optical conversion problems. *thumbsup*

abcdefg30 commented:
Rebased again.

> Erm, can't make out any changes since last time o.O

I squashed the changes into existing commits. I just added https://github.com/OpenRA/OpenRA/pull/12784/commits/80f4feeab30648148e43de75a26bf3d620576781#diff-54b547985b15dbd63a5da7db1eb43a54R422 (the tile 579 is part of the template 101) and https://github.com/OpenRA/OpenRA/pull/12784/commits/80f4feeab30648148e43de75a26bf3d620576781#diff-54b547985b15dbd63a5da7db1eb43a54R413 (tile 342 is part of template 220).

reaperrr commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d4d278844878acd3ee92da7a9c5665e0ab263ec2...9bc45e9c364cf7fea6c17c43071e29cbaa5d1ab8)
